File: lambda_functions/osquery_date_to_elastic_lambda/lambda_function.py
import json
import base64
from datetime import datetime


def date_convert(osquery_date_string):
    old_date = osquery_date_string.replace(" UTC", "")
    old_date_format = "%a %b %d %H:%M:%S %Y"
    new_date = datetime.strptime(old_date, old_date_format)
    new_date_string = new_date.strftime("%Y-%m-%dT%H:%M:%S")
    return new_date_string


def lambda_handler(event, context):
    output = []
    succeeded = 0
    failed = 0
    for record in event['records']:
        #print("original record {}".format(record))
        payload=base64.b64decode(record['data'])
        p = json.loads(payload.decode('utf-8'))
        calendar_time = date_convert(p['calendarTime'])
        if calendar_time:
            p['calendarTime'] = calendar_time
            s = json.dumps(p)
            sb = s.encode('utf-8')
            b64 = str(base64.b64encode(sb))
            output_rec = {
                'recordId': record['recordId'],
                'result': 'Ok',
                'data': base64.standard_b64encode(json.dumps(p))
            }
            succeeded += 1
            #print(output_rec)
        else:
            output_rec = {
                'recordId': record['recordId'],
                'result': 'ProcessingFailed',
                'data': record['data']
            }
            failed += 1
        output.append(output_rec)
    print('Processing completed.  Successful records {}, Failed records {}.'.format(succeeded,
                                                                                    failed))
    return {'records': output}

File: lambda_functions/osquery_date_to_elastic_lambda/lambda_function.go
package main

import (
	"encoding/json"
	"fmt"
	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-lambda-go/lambda"
	"strings"
	"time"
)

//convertData reformats an osquery date stamp to an ElasticSearch readable Timestamp
func convertData(date string) (string, error) {

	oldDate := strings.TrimRight(date, " UTC")
	t, err := time.Parse(time.ANSIC, oldDate)
	if err != nil {
		return "", err
	}

	// Return the properly formatted date
	return t.Format("2006-01-02 15:04:05"), nil
}

//transform recieves a KinesisFirehoseEvent Record
// converts the timestamp to ElasticSearch readable
// and returns a []byte representation of the record data
func transform(rec events.KinesisFirehoseEventRecord) ([]byte, error) {
	recString := string(rec.Data)

	var data map[string]interface{}
	dec := json.NewDecoder(strings.NewReader(recString))
	err := dec.Decode(&data)
	if err != nil {
		return rec.Data, err
	}

	calTime := data["calendarTime"].(string)
	newCalTime, err := convertData(calTime)
	if err != nil {
		return rec.Data, err
	}

	data["calendarTime"] = newCalTime

	jsonData, err := json.Marshal(data)
	if err != nil {
		return rec.Data, err
	}

	return jsonData, nil
}

// Handler is the main AWS Lambda entry point
func Handler(event events.KinesisFirehoseEvent) (events.KinesisFirehoseResponse, error) {
	var results events.KinesisFirehoseResponse

	var failed int

	for _, record := range event.Records {

		responseRec := events.KinesisFirehoseResponseRecord{}
		tranformedData, err := transform(record)

		if err != nil {
			responseRec.RecordID = record.RecordID
			//responseRec.Result = events.KinesisFirehoseTransformedStateProcessingFailed
			responseRec.Result = "ProcessingFailed"
			failed++

		} else {

			responseRec.RecordID = record.RecordID
			//responseRec.Result = events.KinesisFirehoseTransformedStateOk
			responseRec.Result = "Ok"
			responseRec.Data = tranformedData
		}

		results.Records = append(results.Records, responseRec)
	}

	fmt.Printf("Processing completed. Successful records %d, Failed records %d.\n",
		len(results.Records)-failed, failed)

	return results, nil
}

func main() {
	lambda.Start(Handler)
}
